### PR TITLE
Update Leaf Device E2E with scope capabilities

### PR DESCRIPTION
**Pull request by mrohera**
Changes deal with adding capabilities to allow leaf devices to be the in or out of the Edge device's scope. For SAS auth being in scope is optional but it is required for X.509 auth. Any X.509 auth based device not in scope will not be allowed to authenticate with the Edge.

Additional changes are mostly refactoring for better readability, usage and maintenance.

Note:
* Currently C# client and service SDK revisions used are preview as they have not been formally released.
* Warnings as errors was disabled for the LeafDevice project on account of an issue with the preview SDK. Once the SDK is patched and released, this flag can be re enabled.

### CODE REVIEW

File: Azure.IoT.Edge.DotSettings [hunk 1]
```diff
@@ -834,7 +834,8 @@
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ECSharpFileLayoutPatternsUpgrade/@EntryIndexedValue">True</s:Boolean>
     <s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ECSharpKeepExistingMigration/@EntryIndexedValue">True</s:Boolean>
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ECSharpPlaceEmbeddedOnSameLineMigration/@EntryIndexedValue">True</s:Boolean>
-	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ECSharpRenamePlacementToArrangementMigration/@EntryIndexedValue">True</s:Boolean>	
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ECSharpRenamePlacementToArrangementMigration/@EntryIndexedValue">True</s:Boolean>
```
[the remark this answers is not in the document]
Comment (mrohera) [reply Azure.IoT.Edge.DotSettings]: After running the resharper CLI this file was modified, I assumed this should be checked in. Please let me know if it shouldn't.

[the remark this answers is not in the document]
Comment (varunpuranik) [reply Azure.IoT.Edge.DotSettings]: yeah, these have been carefully chosen settings... so I would say please revert this file.

[the remark this answers is not in the document]
Comment (mrohera) [reply Azure.IoT.Edge.DotSettings]: Okay will do.

File: smoke/LeafDevice/LeafDevice.cs [hunk 2]
```diff
@@ -9,65 +9,26 @@ namespace LeafDevice
...
             string eventhubCompatibleEndpointWithEntityPath,
             string deviceId,
             string trustedCACertificateFileName,
             string edgeHostName,
+            string edgeDeviceId,
             bool useWebSockets,
-            string clientCertificatePath,
-            string clientCertificateKeyPath)
+            Option<DeviceCertificate> deviceCertificate,
+            Option<IList<string>> thumprintCertificates)
```
Comment (philipktlin) [inline smoke/LeafDevice/LeafDevice.cs]: typo "thumbprintCertificates"

File: smoke/LeafDevice/LeafDevice.csproj [hunk 3]
```diff
@@ -3,7 +3,7 @@
   <PropertyGroup>
     <OutputType>Exe</OutputType>
     <TargetFramework>netcoreapp2.1</TargetFramework>
-    <TreatWarningsAsErrors>True</TreatWarningsAsErrors>
+    <TreatWarningsAsErrors>False</TreatWarningsAsErrors>
```
Comment (philipktlin) [inline smoke/LeafDevice/LeafDevice.csproj]: Why change this flag? What warnings shown up as errors during build? can we resolve those warnings?

File: smoke/LeafDevice/Program.cs [hunk 4]
```diff
@@ -85,62 +88,55 @@ async Task<int> OnExecuteAsync()
                 string endpoint = this.EventHubCompatibleEndpointWithEntityPath ??
                                   await SecretsHelper.GetSecretFromConfigKey("eventHubConnStrKey");
 
+                var testBuilder = new LeafDevice.Builder(
```
Comment (philipktlin) [inline smoke/LeafDevice/Program.cs]: should it be better to name as "builder"?

File: smoke/LeafDevice/LeafDevice.csproj (hunk 3 above)
Comment (mrohera) [reply smoke/LeafDevice/LeafDevice.csproj]: The warnings were the FxCop style warnings which were part of the SDK.

Comment (mrohera) [reply smoke/LeafDevice/LeafDevice.csproj]: CA2007 is the warning that seems to be coming from the FxCop settings in the SDK.

Comment (philipktlin) [reply smoke/LeafDevice/LeafDevice.csproj]: Remember to set back to True when merging back to master.

Comment (mrohera) [reply smoke/LeafDevice/LeafDevice.csproj]: As noted in the description: Warnings as errors was disabled for the LeafDevice project on account of an issue with the preview SDK. Once the SDK is patched and released, this flag can be re enabled.

**Pull request merged**